velero-plugins/build/restore.go

```diff
@@ -47,6 +48,48 @@ func (p *RestorePlugin) Execute(input *velero.RestoreItemActionExecuteInput) (*v
 	p.Log.Info(fmt.Sprintf("Found new dockercfg secret: %v", secret))
 	build = createNewPushSecret(build, secret)
 
+	registry := build.Annotations[common.RestoreRegistryHostname]
+	if registry == "" {
+		err = fmt.Errorf("failed to find restore registry annotation")
+		return nil, err
+	}
+	// Skip if not internal build
+	name := build.Spec.Strategy.SourceStrategy.From.Name
+	if !strings.HasPrefix(name, build.Annotations[common.BackupRegistryHostname]) {
+		// Does not have internal registry hostname, skip
+		err = fmt.Errorf("build is not from internal image, skipping")
+		return nil, err
+	}
+	shaSplit := strings.Split(name, "@")
+	if len(shaSplit) < 2 {
+		err = fmt.Errorf("unexpected image reference [%v]", name)
+		return nil, err
+	}
+	sha := shaSplit[1]
+	splitName := strings.Split(shaSplit[0], "/")
+	if len(splitName) < 2 {
+		err = fmt.Errorf("unexpected image reference [%v]", name)
+		return nil, err
+	}
+	namespacedName := splitName[len(splitName)-2:]
+	var newName string
+	if namespacedName[0] == "openshift" {
+		// This is a default imagestream/image from existing cluster
+		// Do NOT assume the same sha exists and use floating tag instead
+		newName = fmt.Sprintf("%s/%s/%s", registry, namespacedName[0], namespacedName[1])
+	} else {
+		newName = fmt.Sprintf("%s/%s/%s@%s", registry, namespacedName[0], namespacedName[1], sha)
```

Comment:
Probably not for this PR, but at some point after my PR which includes common/util.go, we should probably switch to using `ReplaceImageRefPrefix` for swapping image references. We'll need to do the same in a number of places.